### PR TITLE
Implement admin control for core modules

### LINKED ISSUE #138: Restrict access to 'admin' modules
As promised, @mrkno, opening up an issue to discuss how you'd imagine approaching the restriction of access to certain core modules/commands that are administrative in nature, e.g.:

- `/disable`
- `/kpm`
- `/restart`
- `/shutdown`
- `/update`

If you have thoughts on how to approach this, that would be great to hear. :-)

### PR DESCRIPTION
**Pull request by mrkno**
resolves #138

### CODE REVIEW

File: core/core_modules/admin.js [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+                if (newArr.length === exports.config.modules[name].length) {
+                    return false;
+                }
+                exports.config.modules[name] = newArr;
+                return true;
+            }
+    }
+},
+
+matchHook = function (moduleName, origionalMatch, config) {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 0.](https://www.codacy.com/app/theprojectshop/Kassy/file/2556616938/issues/source?bid=3282987&fileBranchId=3292072#l98)

File: core/core_modules/admin.js [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    return function (event, commandPrefix) {
+        if (getHasPermission(config, event.sender_id, event.sender_name, event.thread_id, moduleName)) {
+            return origionalMatch(event, commandPrefix);
+        }
+        console.log('no permission')
+        return false;
+    }
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/theprojectshop/Kassy/file/2556616938/issues/source?bid=3282987&fileBranchId=3292072#l105)

File: core/core_modules/admin.js [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        exports.config.modules = {};
+    }
+
+    if (!exports.config.modules[name]) {
+        exports.config.modules[name] = [];
+    }
+
+    switch (action) {
+        case 'create':
+            {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 8 space characters but found 12.](https://www.codacy.com/app/theprojectshop/Kassy/file/2556616938/issues/source?bid=3282987&fileBranchId=3292072#l77)

File: core/core_modules/admin.js [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    if (!exports.config.modules) {
+        exports.config.modules = {};
+    }
+
+    if (!exports.config.modules[name]) {
+        exports.config.modules[name] = [];
+    }
+
+    switch (action) {
+        case 'create':
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 8.](https://www.codacy.com/app/theprojectshop/Kassy/file/2556616938/issues/source?bid=3282987&fileBranchId=3292072#l76)

File: core/core_modules/admin.js [hunk 5]
```diff
@@ -0,0 +1,170 @@
+﻿var config = require.once('../config.js'),
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Irregular whitespace not allowed](https://www.codacy.com/app/theprojectshop/Kassy/file/2556616938/issues/source?bid=3282987&fileBranchId=3292072#l1)

File: core/core_modules/admin.js [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+            var newArr = exports.config.users[name][threadId].filter(function(item) {
+                return item !== action;
+            });
+            if (newArr.length === exports.config.users[name][threadId].length) {
+                return false;
+            }
+            exports.config.users[name][threadId] = newArr;
+            return true;
+        }
+        default:
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 8.](https://www.codacy.com/app/theprojectshop/Kassy/file/2556616938/issues/source?bid=3282987&fileBranchId=3292072#l59)

File: core/core_modules/admin.js [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+        else {
+            api.sendMessage('Complete.', event.thread_id);
+        }
+    }
+    else {
+        api.sendMessage('Only grant, revoke, create and delete are avalible.', event.thread_id);
+    }
+    
+    return false;
+};
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Newline required at end of file but not found.](https://www.codacy.com/app/theprojectshop/Kassy/file/2556616938/issues/source?bid=3282987&fileBranchId=3292072#l170)

File: core/core_modules/admin.js [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    var common = user[threadId].filter(function (p) {
+        return config.modules[moduleName].indexOf(p) !== -1;
+    });
+
+    return common.length >= 1;
+},
+
+modify = function (name, threadId, action, method) {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 0.](https://www.codacy.com/app/theprojectshop/Kassy/file/2556616938/issues/source?bid=3282987&fileBranchId=3292072#l24)

File: core/core_modules/admin.js [hunk 9]
```diff
@@ -0,0 +1,170 @@
+﻿var config = require.once('../config.js'),
+
+getHasPermission = function (config, userId, name, threadId, moduleName) {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 0.](https://www.codacy.com/app/theprojectshop/Kassy/file/2556616938/issues/source?bid=3282987&fileBranchId=3292072#l3)

File: core/core_modules/admin.js [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+            }
+    }
+},
+
+matchHook = function (moduleName, origionalMatch, config) {
+    return function (event, commandPrefix) {
+        if (getHasPermission(config, event.sender_id, event.sender_name, event.thread_id, moduleName)) {
+            return origionalMatch(event, commandPrefix);
+        }
+        console.log('no permission')
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/theprojectshop/Kassy/file/2556616938/issues/source?bid=3282987&fileBranchId=3292072#l103)

File: core/core_modules/admin.js [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+};
+
+exports.match = function (event, commandPrefix) {
+    return event.arguments[0] === commandPrefix + 'admin';
+};
+
+exports.help = function (commandPrefix) {
+    return [
+        [commandPrefix + 'admin <grant/revoke> <fullName/userId> <permissionName>', 'Grant or revoke a permission for a user.',
+            'Assigns/removes a permission from a user. Once a user has a permission they can use any core modules with the corresponding permission (and any that have no permissions/none configured).'],
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Line is too long.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560099488/issues/source?bid=3282987&fileBranchId=3292072#l128)

File: core/core_modules/admin.js [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+    }
+},
+
+matchHook = function (moduleName, origionalMatch, config) {
+    return function (event, commandPrefix) {
+        if (getHasPermission(config, event.sender_id, event.sender_name, event.thread_id, moduleName)) {
+            return origionalMatch(event, commandPrefix);
+        }
+        return false;
+    }
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560099488/issues/source?bid=3282987&fileBranchId=3292072#l98)

File: core/core_modules/admin.js [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+            if (newArr.length === exports.config.users[name][threadId].length) {
+                return false;
+            }
+            exports.config.users[name][threadId] = newArr;
+            return true;
+        }
+    }
+},
+
+setup = function (action, name, permission) {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 0.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560099488/issues/source?bid=3282987&fileBranchId=3292072#l60)

File: core/core_modules/admin.js [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+    switch (action) {
+        case 'create':
+            {
+                if (!exports.config.modules[name].includes(permission)) {
+                    exports.config.modules[name].push(permission);
+                    return true;
+                }
+                return false;
+            }
+        case 'delete':
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 8.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560099488/issues/source?bid=3282987&fileBranchId=3292072#l78)

File: core/core_modules/admin.js (hunk 12 above)
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560099488/issues/source?bid=3282987&fileBranchId=3292072#l98)

File: core/core_modules/admin.js [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+    if (!exports.config.modules) {
+        exports.config.modules = {};
+    }
+
+    if (!exports.config.modules[name]) {
+        exports.config.modules[name] = [];
+    }
+
+    switch (action) {
+        case 'create':
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 8.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560099488/issues/source?bid=3282987&fileBranchId=3292072#l70)

File: core/core_modules/admin.js [hunk 16]
```diff
@@ -0,0 +1,167 @@
+﻿var config = require.once('../config.js'),
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Irregular whitespace not allowed](https://www.codacy.com/app/theprojectshop/Kassy/file/2560099488/issues/source?bid=3282987&fileBranchId=3292072#l1)

File: core/core_modules/admin.js [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    var common = user[threadId].filter(function (p) {
+        return config.modules[moduleName].indexOf(p) !== -1;
+    });
+
+    return common.length >= 1;
+},
+
+modify = function (name, threadId, action, method) {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 0.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560099488/issues/source?bid=3282987&fileBranchId=3292072#l24)

File: core/core_modules/admin.js [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+        exports.config.modules = {};
+    }
+
+    if (!exports.config.modules[name]) {
+        exports.config.modules[name] = [];
+    }
+
+    switch (action) {
+        case 'create':
+            {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 8 space characters but found 12.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560099488/issues/source?bid=3282987&fileBranchId=3292072#l71)

File: core/core_modules/admin.js [hunk 19]
```diff
@@ -0,0 +1,167 @@
+﻿var config = require.once('../config.js'),
+
+getHasPermission = function (config, userId, name, threadId, moduleName) {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 0.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560099488/issues/source?bid=3282987&fileBranchId=3292072#l3)

File: core/core_modules/admin.js [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+};
+
+exports.match = function (event, commandPrefix) {
+    return event.arguments[0] === commandPrefix + 'admin';
+};
+
+exports.help = function (commandPrefix) {
+    return [
+        [commandPrefix + 'admin <grant/revoke> <fullName/userId> <permissionName>', 'Grant or revoke a permission for a user.',
+            'Assigns/removes a permission from a user. Once a user has a permission they can use any core modules with the corresponding permission (and any that have no permissions/none configured).'],
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Line is too long.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560344994/issues/source?bid=3282987&fileBranchId=3292072#l128)

File: core/core_modules/admin.js [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+		if (!exports.config.users) {
+			exports.config.users = {};
+		}
+
+		if (!exports.config.users[name]) {
+			exports.config.users[name] = {};
+		}
+
+		if (!exports.config.users[name][threadId]) {
+			exports.config.users[name][threadId] = [];
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 0.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560344994/issues/source?bid=3282987&fileBranchId=3292072#l34)

File: core/core_modules/admin.js [hunk 22]
```diff
@@ -0,0 +1,167 @@
...
+		}
+
+		switch (method) {
+			case 'grant':
+			{
+				if (!exports.config.users[name][threadId].includes(action)) {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 0.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560344994/issues/source?bid=3282987&fileBranchId=3292072#l40)

File: core/core_modules/admin.js [hunk 23]
```diff
@@ -0,0 +1,167 @@
...
+		if (!config.modules || !config.modules[moduleName] || config.modules[moduleName].length === 0) {
+			return true;
+		}
+
+		if (!config.users || Object.keys(config.users).length === 0) {
+			return false;
+		}
+
+		var user = config.users[userId] || config.users[name];
+		if (!user || !user[threadId] || user[threadId].length === 0) {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 0.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560344994/issues/source?bid=3282987&fileBranchId=3292072#l13)

File: core/core_modules/admin.js [hunk 24]
```diff
@@ -0,0 +1,167 @@
...
+
+		if (!exports.config.modules[name]) {
+			exports.config.modules[name] = [];
+		}
+
+		switch (action) {
+			case 'create':
+				{
+					if (!exports.config.modules[name].includes(permission)) {
+						exports.config.modules[name].push(permission);
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 0.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560344994/issues/source?bid=3282987&fileBranchId=3292072#l73)

File: core/core_modules/admin.js [hunk 25]
```diff
@@ -0,0 +1,167 @@
...
+				if (newArr.length === exports.config.users[name][threadId].length) {
+					return false;
+				}
+				exports.config.users[name][threadId] = newArr;
+				return true;
+			}
+		}
+	},
+
+	setup = function (action, name, permission) {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 0.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560344994/issues/source?bid=3282987&fileBranchId=3292072#l60)

File: core/core_modules/admin.js [hunk 26]
```diff
@@ -0,0 +1,167 @@
...
+					return true;
+				}
+				return false;
+			}
+			case 'revoke':
+			{
+				var newArr = exports.config.users[name][threadId].filter(function(item) {
+					return item !== action;
+				});
+				if (newArr.length === exports.config.users[name][threadId].length) {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 0.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560344994/issues/source?bid=3282987&fileBranchId=3292072#l51)

File: core/core_modules/admin.js [hunk 27]
```diff
@@ -0,0 +1,167 @@
...
+		var user = config.users[userId] || config.users[name];
+		if (!user || !user[threadId] || user[threadId].length === 0) {
+			return false;
+		}
+
+		var common = user[threadId].filter(function (p) {
+			return config.modules[moduleName].indexOf(p) !== -1;
+		});
+
+		return common.length >= 1;
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 0.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560344994/issues/source?bid=3282987&fileBranchId=3292072#l21)

File: core/core_modules/admin.js [hunk 28]
```diff
@@ -0,0 +1,167 @@
...
+	modify = function (name, threadId, action, method) {
+		if (!exports.config.users) {
+			exports.config.users = {};
+		}
+
+		if (!exports.config.users[name]) {
+			exports.config.users[name] = {};
+		}
+
+		if (!exports.config.users[name][threadId]) {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 0.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560344994/issues/source?bid=3282987&fileBranchId=3292072#l33)

File: core/core_modules/admin.js [hunk 29]
```diff
@@ -0,0 +1,167 @@
...
+			}
+			case 'revoke':
+			{
+				var newArr = exports.config.users[name][threadId].filter(function(item) {
+					return item !== action;
+				});
+				if (newArr.length === exports.config.users[name][threadId].length) {
+					return false;
+				}
+				exports.config.users[name][threadId] = newArr;
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 0.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560344994/issues/source?bid=3282987&fileBranchId=3292072#l54)

File: core/core_modules/admin.js [hunk 30]
```diff
@@ -0,0 +1,163 @@
...
+};
+
+exports.match = function (event, commandPrefix) {
+    return event.arguments[0] === commandPrefix + 'admin';
+};
+
+exports.help = function (commandPrefix) {
+    return [
+        [commandPrefix + 'admin <grant/revoke> <fullName/userId> <permissionName>', 'Grant or revoke a permission for a user.',
+            'Assigns/removes a permission from a user. Once a user has a permission they can use any core modules with the corresponding permission (and any that have no permissions/none configured).'],
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Line is too long.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560420210/issues/source?bid=3282987&fileBranchId=3292072#l124)

File: core/core_modules/admin.js [hunk 31]
```diff
@@ -0,0 +1,163 @@
...
+            exports.config.modules = {};
+        }
+
+        if (!exports.config.modules[name]) {
+            exports.config.modules[name] = [];
+        }
+
+        switch (action) {
+			case 'create': {
+                if (!exports.config.modules[name].includes(permission)) {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 0.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560420210/issues/source?bid=3282987&fileBranchId=3292072#l69)

File: core/core_modules/admin.js [hunk 32]
```diff
@@ -0,0 +1,163 @@
...
+                    exports.config.modules[name].push(permission);
+                    return true;
+                }
+                return false;
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 0.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560420210/issues/source?bid=3282987&fileBranchId=3292072#l73)

File: core/core_modules/admin.js [hunk 33]
```diff
@@ -0,0 +1,163 @@
...
+        if (!exports.config.users[name]) {
+            exports.config.users[name] = {};
+        }
+
+        if (!exports.config.users[name][threadId]) {
+            exports.config.users[name][threadId] = [];
+        }
+
+        switch (method) {
+            case 'grant': {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 8 space characters but found 12.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560420210/issues/source?bid=3282987&fileBranchId=3292072#l38)

File: core/core_modules/admin.js [hunk 34]
```diff
@@ -0,0 +1,163 @@
+﻿var config = require.once('../config.js'),
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Irregular whitespace not allowed](https://www.codacy.com/app/theprojectshop/Kassy/file/2560420210/issues/source?bid=3282987&fileBranchId=3292072#l1)

File: core/core_modules/admin.js [hunk 35]
```diff
@@ -0,0 +1,163 @@
...
+        }
+
+        switch (action) {
+			case 'create': {
+                if (!exports.config.modules[name].includes(permission)) {
+                    exports.config.modules[name].push(permission);
+                    return true;
+                }
+                return false;
+            }
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 0 space characters but found 12.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560420210/issues/source?bid=3282987&fileBranchId=3292072#l74)

File: core/core_modules/admin.js [hunk 36]
```diff
@@ -0,0 +1,163 @@
...
+        if (!exports.config.modules) {
+            exports.config.modules = {};
+        }
+
+        if (!exports.config.modules[name]) {
+            exports.config.modules[name] = [];
+        }
+
+        switch (action) {
+			case 'create': {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 8 space characters but found 0.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560420210/issues/source?bid=3282987&fileBranchId=3292072#l68)

File: core/core_modules/admin.js [hunk 37]
```diff
@@ -0,0 +1,163 @@
...
+
+        switch (method) {
+            case 'grant': {
+                if (!exports.config.users[name][threadId].includes(action)) {
+                    exports.config.users[name][threadId].push(action);
+                    return true;
+                }
+                return false;
+            }
+            case 'revoke': {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 8 space characters but found 12.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560420210/issues/source?bid=3282987&fileBranchId=3292072#l45)

File: core/core_modules/admin.js [hunk 38]
```diff
@@ -0,0 +1,163 @@
...
+
+        switch (action) {
+			case 'create': {
+                if (!exports.config.modules[name].includes(permission)) {
+                    exports.config.modules[name].push(permission);
+                    return true;
+                }
+                return false;
+            }
+            case 'delete': {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 8 space characters but found 12.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560420210/issues/source?bid=3282987&fileBranchId=3292072#l75)

File: core/core_modules/admin.js [hunk 39]
```diff
@@ -0,0 +1,164 @@
...
+        switch (method) {
+        case 'grant': {
+                if (!exports.config.users[name][threadId].includes(action)) {
+                    exports.config.users[name][threadId].push(action);
+                    return true;
+                }
+                return false;
+            }
+        case 'revoke': {
+                var newArr = exports.config.users[name][threadId].filter(function(item) {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 0.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560513839/issues/source?bid=3282987&fileBranchId=3292072#l46)

File: core/core_modules/admin.js [hunk 40]
```diff
@@ -0,0 +1,164 @@
...
+                    return item !== action;
+                });
+                if (newArr.length === exports.config.users[name][threadId].length) {
+                    return false;
+                }
+                exports.config.users[name][threadId] = newArr;
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 0.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560513839/issues/source?bid=3282987&fileBranchId=3292072#l52)

File: core/core_modules/admin.js [hunk 41]
```diff
@@ -0,0 +1,164 @@
...
+                    exports.config.users[name][threadId].push(action);
+                    return true;
+                }
+                return false;
+            }
+        case 'revoke': {
+                var newArr = exports.config.users[name][threadId].filter(function(item) {
+                    return item !== action;
+                });
+                if (newArr.length === exports.config.users[name][threadId].length) {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 0.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560513839/issues/source?bid=3282987&fileBranchId=3292072#l49)

File: core/core_modules/admin.js [hunk 42]
```diff
@@ -0,0 +1,164 @@
+﻿var config = require.once('../config.js'),
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Irregular whitespace not allowed](https://www.codacy.com/app/theprojectshop/Kassy/file/2560513839/issues/source?bid=3282987&fileBranchId=3292072#l1)

File: core/core_modules/admin.js [hunk 43]
```diff
@@ -0,0 +1,164 @@
...
+            }
+        case 'revoke': {
+                var newArr = exports.config.users[name][threadId].filter(function(item) {
+                    return item !== action;
+                });
+                if (newArr.length === exports.config.users[name][threadId].length) {
+                    return false;
+                }
+                exports.config.users[name][threadId] = newArr;
+                return true;
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 0.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560513839/issues/source?bid=3282987&fileBranchId=3292072#l53)

File: core/core_modules/admin.js [hunk 44]
```diff
@@ -0,0 +1,164 @@
...
+        }
+
+        switch (method) {
+        case 'grant': {
+                if (!exports.config.users[name][threadId].includes(action)) {
+                    exports.config.users[name][threadId].push(action);
+                    return true;
+                }
+                return false;
+            }
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 0 space characters but found 12.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560513839/issues/source?bid=3282987&fileBranchId=3292072#l44)

File: core/core_modules/admin.js [hunk 45]
```diff
@@ -0,0 +1,164 @@
...
+            exports.config.users[name][threadId] = [];
+        }
+
+        switch (method) {
+        case 'grant': {
+                if (!exports.config.users[name][threadId].includes(action)) {
+                    exports.config.users[name][threadId].push(action);
+                    return true;
+                }
+                return false;
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 0.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560513839/issues/source?bid=3282987&fileBranchId=3292072#l43)

File: core/core_modules/admin.js [hunk 46]
```diff
@@ -0,0 +1,164 @@
...
+            exports.config.users[name] = {};
+        }
+
+        if (!exports.config.users[name][threadId]) {
+            exports.config.users[name][threadId] = [];
+        }
+
+        switch (method) {
+        case 'grant': {
+                if (!exports.config.users[name][threadId].includes(action)) {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 4 space characters but found 0.](https://www.codacy.com/app/theprojectshop/Kassy/file/2560513839/issues/source?bid=3282987&fileBranchId=3292072#l39)

**Comment by mrkno** [comment]
Codacy is drunk. Half the things it is complaining about are on the old diff or I cannot determine why.

**Comment by MatthewHartstonge** [comment]
Did you merge a PR? It might run all open PRs against master again.

**Comment by mrkno** [comment]
@MatthewHartstonge currently not merged, cause have not had confirmation from @Awarua- if he thinks it is ok.

**Comment by Awarua-** [comment]
Would it be a good idea for the module to support lists of users and core modules, as in you could create permission groups on a list of modules, in one call instead of many?

**Comment by mrkno** [comment]
The primary idea is that you would manually edit the config file, rather than doing this through the commands - there is no good way to provide two variable length lists of items within the same command.

The commands themselves are only if you want to adjusst the permissions at runtime.

i.e. use case might be adding a new admin or removing an old one, not creating the entire permissions structure from scratch

File: core/core_modules/admin.js [hunk 47]
```diff
@@ -0,0 +1,164 @@
...
+        if (!exports.config.users[name]) {
+            exports.config.users[name] = {};
+        }
+
+        if (!exports.config.users[name][threadId]) {
+            exports.config.users[name][threadId] = [];
+        }
+
+        switch (method) {
+            case 'grant': {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 8 space characters but found 12.](https://www.codacy.com/app/theprojectshop/Kassy/file/2571810516/issues/source?bid=3282987&fileBranchId=3292072#l38)

File: core/core_modules/admin.js (hunk 42 above)
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Irregular whitespace not allowed](https://www.codacy.com/app/theprojectshop/Kassy/file/2571810516/issues/source?bid=3282987&fileBranchId=3292072#l1)

File: core/core_modules/admin.js [hunk 48]
```diff
@@ -0,0 +1,164 @@
...
+                if (!exports.config.users[name][threadId].includes(action)) {
+                    exports.config.users[name][threadId].push(action);
+                    return true;
+                }
+                return false;
+            }
+            case 'revoke': {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 8 space characters but found 12.](https://www.codacy.com/app/theprojectshop/Kassy/file/2571810516/issues/source?bid=3282987&fileBranchId=3292072#l45)

File: core/core_modules/admin.js [hunk 49]
```diff
@@ -0,0 +1,176 @@
+﻿var config = require.once('../config.js'),
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Irregular whitespace not allowed](https://www.codacy.com/app/theprojectshop/Kassy/file/2573600580/issues/source?bid=3282987&fileBranchId=3292072#l1)

File: core/core_modules/admin.js [hunk 50]
```diff
@@ -0,0 +1,176 @@
...
+
+    getHasPermission = function (config, userId, name, threadId, moduleName) {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (12)](https://www.codacy.com/app/theprojectshop/Kassy/file/2573600580/issues/source?bid=3282987&fileBranchId=3292072#l3)

File: core/core_modules/admin.js [hunk 51]
```diff
@@ -0,0 +1,176 @@
...
+
+        switch (method) {
+            case 'grant': {
+                if (!exports.config.users[name][threadId].includes(action)) {
+                    exports.config.users[name][threadId].push(action);
+                    return true;
+                }
+                return false;
+            }
+            case 'revoke': {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 8 space characters but found 12.](https://www.codacy.com/app/theprojectshop/Kassy/file/2573600580/issues/source?bid=3282987&fileBranchId=3292072#l57)

File: core/core_modules/admin.js [hunk 52]
```diff
@@ -0,0 +1,176 @@
...
+        if (!exports.config.users[name]) {
+            exports.config.users[name] = {};
+        }
+
+        if (!exports.config.users[name][threadId]) {
+            exports.config.users[name][threadId] = [];
+        }
+
+        switch (method) {
+            case 'grant': {
```
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 8 space characters but found 12.](https://www.codacy.com/app/theprojectshop/Kassy/file/2573600580/issues/source?bid=3282987&fileBranchId=3292072#l50)

File: core/core_modules/admin.js (hunk 50 above)
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Function 'anonymous' has a complexity of 12.](https://www.codacy.com/app/theprojectshop/Kassy/file/2573600580/issues/source?bid=3282987&fileBranchId=3292072#l3)

File: core/core_modules/admin.js (hunk 49 above)
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Irregular whitespace not allowed](https://www.codacy.com/app/theprojectshop/Kassy/file/2587223488/issues/source?bid=3282987&fileBranchId=3292072#l1)

File: core/core_modules/admin.js (hunk 50 above)
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (12)](https://www.codacy.com/app/theprojectshop/Kassy/file/2587223488/issues/source?bid=3282987&fileBranchId=3292072#l3)

File: core/core_modules/admin.js (hunk 51 above)
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 8 space characters but found 12.](https://www.codacy.com/app/theprojectshop/Kassy/file/2587223488/issues/source?bid=3282987&fileBranchId=3292072#l57)

File: core/core_modules/admin.js (hunk 52 above)
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Expected indentation of 8 space characters but found 12.](https://www.codacy.com/app/theprojectshop/Kassy/file/2587223488/issues/source?bid=3282987&fileBranchId=3292072#l50)

File: core/core_modules/admin.js (hunk 50 above)
Comment (mrkno) [inline core/core_modules/admin.js]: Codacy Issue found: [Function 'anonymous' has a complexity of 12.](https://www.codacy.com/app/theprojectshop/Kassy/file/2587223488/issues/source?bid=3282987&fileBranchId=3292072#l3)

**Comment by Awarua-** [comment]
Codacy is drunk, Approved

**Pull request merged**